README.md
### Libpruio Debian package ###

This package includes libpruio, with some patches applied as well as
addition of files needed for building a debian package.

Testing Debian packages (will be) available at: http://beagle.tuks.nl

Original source:

 https://www.freebasic-portal.de/dlfiles/592/libpruio-0.2.tar.bz2

Github repository:

 https://github.com/l4m4re/libpruio

The initial commit on the repository consists of the original source, so
the repository contains all patches and changes as applied to the
original source. 

Since there are no Makefiles in the original source, I wrote and used a
generic build and install system using gmake, which is available at:

 https://github.com/l4m4re/LaMake
 
--Arend, July 2016

### Original README.txt below ###

This package includes the source code of libpruio, a library supporting
digital and analog input and output capabilities for the Beaglebone
hardware with TI AM3359 CPU. Also examples and a tool for device tree
overlays are included.

(C) 2014 by Thomas{ dOt ]Freiherr[ aT ]gmx[ DoT }net

Library licence:

  LGPL2 (see http://www.gnu.org/licenses/lgpl-2.0.html)

Examples licence:

  GPL3 (see http://www.gnu.org/licenses/gpl-3.0.html)

Find a more detailed description at

  http://beagleboard.org/project/libpruio/
  http://www.freebasic.net/forum/viewtopic.php?f=14&t=22501

Find the on-line documentation at

  http://users.freebasic-portal.de/tjf/Projekte/libpruio/doc/html/index.html

Find the off-line documentation at

  http://www.freebasic-portal.de/downloads/fb-on-arm/anleitung-zu-libpruio-en-326.html

Find installation instructions at

  http://users.freebasic-portal.de/tjf/Projekte/libpruio/doc/html/_cha_preparation.html#SecInstallation


debian/changelog
libpruio (0.2) unstable; urgency=low

  * Initial Release.

 -- Arend Lammertink <lamare@gmail.com>  Sat, 09 Jul 2016 12:57:27 +0000


src/config/mk_dts.bas
/'* \file dts_universal.bas
\brief Tool to create, compile and install an universal device tree overlay for libpruio for run-time pinmuxing.

This is a helper tool for an universal device tree overlay. Adapt this
FB source code, compile it and run the executable. This will create a
device tree overlay source file in the current directory. 

The universal overlay provides pinmuxing capability at run-time. Root
privileges are required to achieve this. It claims all header pins and
prepares configurations for all modes. By default the free header pins
on a Beaglebone Green are declared. Customizing can be done to include
more pins (ie. the HDMI pins when not used) or to reduce the number of
pins (ie. when they interfere with other capes).

- To include a pins group uncomment the matching `PIN_DEL(...)` line, or

- to drop a pin add a code line like `M(P8_08) = ""` right below the
  `PIN_DEL(...)` lines.

Licence: GPLv3

Copyrighe 2014 by Thomas{ dOt ]Freiherr[ At ]gmx[ DoT }net
          2016 by lamare[ At ]gmail[ DoT }com

Adapted from dts_universal.bas by lamare:

- Removed the -00A0 version from the output filename.

- Removed the call to the dts compiler, since that is handled by LaMake.

- Removed the pins for BB-UART2, which is the Grove UART on the
  BeagleBone Green. So, enabling these gives a conflict on the Green, as
  stated by Thomas:

https://groups.google.com/forum/#!msg/beagleboard/CN5qKSmPIbc/2UdstwjNhHMJ 

    You need not disable UART2 for BBG. Instead you can disable libpruio
    control for pin P9_22 in file dts_universal.bas by adding a line

    M(P9_22) = "" '' right below the PIN_DEL(...) lines.

    in order to release this pin from the libpruio overlay. Also, you can
    free the HDMI pins by removing the line

    PIN_DEL(HDMI_Pins)

    Adapt before running the command fbc -w all dts_universal.bas.


\since 0.2
'/

#INCLUDE ONCE "pruiotools.bas"

'''''''''''''''''''''''''''''''''''''''''''''''''''''''' adapt this code

'* The file name.
#DEFINE FILE_NAME "libpruio"
'* The version.
#DEFINE VERS_NAME "00A0"

' quick & dirty: first create settings for all pins ...
#INCLUDE ONCE "P8.bi"
#INCLUDE ONCE "P9.bi"
' ... then delete unwanted
PIN_DEL(HDMI_Pins)
PIN_DEL(EMMC2_Pins)
PIN_DEL(I2C1_Pins)
PIN_DEL(I2C2_Pins)
PIN_DEL(MCASP0_Pins)

' BBG Grove UART:
M(P9_22) = ""

''''''''''''''''''''''''''''''''''''''''''''''''''''''' end of adaptions


VAR fnam = FILE_NAME, _    '*< The file name (without path / suffix)
     fnr = FREEFILE        '*< The file number.
IF OPEN(fnam & ".dts" FOR OUTPUT AS fnr) THEN
'IF OPEN CONS(FOR OUTPUT AS #fnr) THEN
  '?"failed openig console"
  ?"failed writing file: " & fnam & ".dts"
ELSE
  PRINT #fnr, ALL_START;

  FOR i AS LONG = 0 TO UBOUND(M)
    VAR x = IIF(LEN(M(i)), nameBall(i), 0) '*< The header pin name.
    IF x THEN PRINT #fnr, ENTRY_EXCL(*x);
  NEXT

  PRINT #fnr, FRAG0_START;

  FOR i AS LONG = 0 TO UBOUND(M)
    IF LEN(M(i)) THEN PRINT #fnr, f0entry(i);
  NEXT

  PRINT #fnr, FRAG0_END;
  PRINT #fnr, FRAG1_START;

  FOR i AS LONG = 0 TO UBOUND(M)
    IF LEN(M(i)) THEN PRINT #fnr, f1entry(i);
  NEXT

  PRINT #fnr, FRAG1_END;
  PRINT #fnr, ALL_END;
  CLOSE #fnr

END IF


src/c_examples/rb_file.c
/* Sources:

https://groups.google.com/forum/#!msg/beagleboard/kxxucJAci2c/IDF7nCdRJgAJ

http://www.freebasic.net/forum/viewtopic.php?t=22501&start=105



*/


/** \file rb_file.c
\brief Example: fetch ADC samples in a ring buffer and save to file.
 
This file contains an example on how to use the ring buffer mode of
libpruio. A fixed step mask of AIN-0, AIN-1 and AIN-2 get configured
for maximum speed, sampled in to the ring buffer and from there saved
as raw data to some files.
 
Licence: GPLv3
 
Copyright 2014-2015 by Thomas{ dOt ]Freiherr[ At ]gmx[ DoT }net
 
Thanks for C code translation: Nils Kohrs (nils.kohrs [at] gmail.com)
 
Compile by: `gcc -Wall -o rb_file rb_file.c -lpruio -lprussdrv`
 
\since 0.2.0.2
*/
 
#include "unistd.h"
#include "time.h"
#include "stdio.h"
//#include "../c_include/pruio.h"
#include "../c_wrapper/pruio.h"
//#include "../c_wrapper/pruio_pins.h"
 
//! The main function.
int main(int argc, char **argv)
{
  const uint32 tSamp = 123401;  //!< The number of samples in the files (per step).
  const uint32 tmr = 5000;      //!< The sampling rate in ns (5000 -> 200 kHz).
  const uint32 NoStep = 3;      //!< The number of active steps (must match setStep calls and mask).
  const uint32 NoFile = 2;      //!< The number of files to write.
  const char *NamFil = "output.%u"; //!< The output file names.
  struct timespec mSec;
  mSec.tv_nsec=1000000;
  pruIo *io = pruio_new(PRUIO_DEF_ACTIVE, 0x98, 0, 1); //! create new driver
  if (io->Errr){
               printf("constructor failed (%s)\n", io->Errr); return 1;}
 
  do {
    if (pruio_adc_setStep(io, 9, 0, 0, 0, 0)){ //          step 9, AIN-0
        printf("step 9 configuration failed: (%s)\n", io->Errr); break;}
    if (pruio_adc_setStep(io,10, 1, 0, 0, 0)){ //         step 10, AIN-1
       printf("step 10 configuration failed: (%s)\n", io->Errr); break;}
    if (pruio_adc_setStep(io,11, 2, 0, 0, 0)){ //         step 11, AIN-2
       printf("step 11 configuration failed: (%s)\n", io->Errr); break;}
 
    uint32 mask = 7 << 9;         //!< The active steps (9 to 11).
    uint32 tInd = tSamp * NoStep; //!< The maximum total index.
    uint32 half = ((io->ESize >> 2) / NoStep) * NoStep; //!< The maximum index of the half ring buffer.
 
    if (half > tInd){ half = tInd;}  //       adapt size for small files
    uint32 samp = (half << 1) / NoStep; //!< The number of samples (per step).
 
    if (pruio_config(io, samp, mask, tmr, 0)){ //       configure driver
                       printf("config failed (%s)\n", io->Errr); break;}
 
    if (pruio_rb_start(io)){
                     printf("rb_start failed (%s)\n", io->Errr); break;}
 
    uint16 *p0 = io->Adc->Value;  //!< A pointer to the start of the ring buffer.
    uint16 *p1 = p0 + half;       //!< A pointer to the middle of the ring buffer.
    uint32 n;  //!< File counter.
    char fName[20];
    for(n = 0; n < NoFile; n++){
      sprintf(fName, NamFil, n);
      printf("Creating file %s\n", fName);
      FILE *oFile = fopen(fName, "wb");
      uint32 i = 0;               //!< Start index.
      while(i < tInd){
        i += half;
        if(i > tInd){             // fetch the rest(no complete chunk)
          uint32 rest = tInd + half - i;
          uint32 iEnd = p1 >= p0 ? rest : rest + half;
          while(io->DRam[0] < iEnd) nanosleep(&mSec, NULL);
          printf("  writing samples %u-%u\n", tInd -rest, tInd-1);
          fwrite(p0, sizeof(uint16), rest, oFile);
          uint16 *swap = p0;
          p0 = p1;
          p1 = swap;
        }
        if(p1 > p0) while(io->DRam[0] < half) nanosleep(&mSec, NULL);
        else        while(io->DRam[0] > half) nanosleep(&mSec, NULL);
        printf("  writing samples %u-%u\n", i-half, i-1);
        fwrite(p0, sizeof(uint16), half, oFile);
        uint16 *swap = p0;
        p0 = p1;
        p1 = swap;
      }
      fclose(oFile);
      printf("Finished file %s\n", fName);
    }
  } while(0);
  pruio_destroy(io);
  return 0;
} 



src/c_examples/rb_plot.c
/** \file rb_plot.c 
\brief Example: fetch ADC samples in a ring buffer, save to file and view by GnuPlot. 
 
This file contains an example on how to use the ring buffer mode of 
libpruio. A fixed step mask of AIN-0 (P9_39) get configured 
for maximum speed, sampled in to the ring buffer and from there saved 
to a files and viewed by GnuPlot
 
Licence: GPLv3 
 
Copyright 2016 by Axel Hülsmann (axhuelsmann[at]analog-photonics[dot]com)

Thanks:		for libpruio: Thomas{ dOt ]Freiherr[ At ]gmx[ DoT }net 
		for C code translation: Nils Kohrs (nils.kohrs [at] gmail.com) 

Prepare: 	sudo su      						(always)
		apt-get install gnuplot					(once)
		echo BB-BONE-PRU-01 > /sys/devices/bone_capemgr.9/slots (after reboot)
 
Compile by: 	gcc -Wall -o rb_plot rb_plot.c -lpruio 			(-lprussdrv)
*/ 
 
#include "stdio.h" 
#include "../c_wrapper/pruio.h" 
 
int main(int argc, char **argv) 
{ 
  const uint16  Act = PRUIO_DEF_ACTIVE; // activation mode
  const uint8    Av = 0;	// avaraging for default steps
  const uint32  OpD = 0;	// open delay for default steps (default 0x98, max 0x3FFFF)
  const uint8   SaD = 0;	// sample delay for default steps (defaults to 0)
  const uint8   Stp = 9;	// step index (0=step 0=>charge step, 1=step 1 (=> AIN-0 by default), ...., 17 = idle step)
  const uint8   ChN = 0;	// Channel Number to scan (0 = AIN-0, 1= AIN-1, ....)
  const uint16  Mds = 4;	// modus for output (default to 4 = 16 bit)
  const uint32 samp = 500;      // number of samples in the files (per step) 
        uint32 mask = 1 << Stp;	// mask for active steps (default to all 8 channels active in steps 1 to 8)
  const uint32  tmr = 10000;	// 1ms! sampling rate in ns (10000 -> 100 kHz) 
  	uint32    k = 0; 	// counter for array
  	uint16 array[samp]; 	// array for text file

  pruIo *io = pruio_new(Act, Av, OpD, SaD); // create new driver
  if (io->Errr){ printf("constructor failed (%s)\n", io->Errr); return 1;} 
 
  if (pruio_adc_setStep(io, Stp, ChN, Av, SaD, OpD)) printf("configuration failed: (%s)\n", io->Errr); 

  if (pruio_config(io, samp, mask, tmr, Mds)) printf("config failed (%s)\n", io->Errr);
 
  if (pruio_rb_start(io))printf("rb_start failed (%s)\n", io->Errr); // start measurement
 
  uint16 *p = io->Adc->Value;  // pointer to start of ring buffer. 

  FILE *plotfile = fopen("output.dat", "w+"); 

  	for( k = 0; k < samp; k++) {
		array[k] = *(p + k); // reading ring buffer
		fprintf(plotfile , "%i \n", array[k]);}

  fclose(plotfile);
  pruio_destroy(io); 

  // view adc in GnuPlot using pipes
  FILE *gnuplotPipe = popen("gnuplot -persistent" , "w");
  fprintf(gnuplotPipe, "plot 'output.dat'\n");
  pclose(gnuplotPipe);
  return 0; 
} 


src/config/pruiotools.bas
/'* \file pruiotools.bas
\brief Common source code for the helper tools dts_custom.bas and dts_universal.bas.

This file contains common source code used by the device tree overlay
helper tools dts_custom.bas and dts_universal.bas. Do not edit (experts
only).

Licence: GPLv3

Copyright 2014 by Thomas{ dOt ]Freiherr[ At ]gmx[ DoT }net

\since 0.2

'/

'* The name of the pinmux folders in /sys/devices/ocp.* (must match the definition in pruio.bas).
#DEFINE PMUX_NAME "pruio-"
TYPE AS UBYTE uint8 '*< Type alias.
' Convenience declarations.
#INCLUDE ONCE "../pruio/pruio_pins.bi"

'* The start of the source file
#DEFINE ALL_START _
  "// dts file auto generated by pruio_config (don't edit)" _
  !"\n/dts-v1/;" _
  !"\n/plugin/;" _
  !"\n" _
  !"\n/ {" _
  !"\n    compatible = ""ti,beaglebone"", ""ti,beaglebone-black"", ""ti,beaglebone-green"" ;" _
  !"\n" _
  !"\n    // identification" _
  !"\n    board-name = """ & FILE_NAME & """;" _
  !"\n    manufacturer = ""TJF"";" _
  !"\n    part-number = ""PruIoBBB"";" _
  !"\n    version = """ & VERS_NAME & """;" _
  !"\n" _
  !"\n    // state the resources this cape uses" _
  !"\n    exclusive-use ="

'* An entry line for the `exclusive-use =` section.
#DEFINE ENTRY_EXCL(_T_) !"\n      """ & _T_ & ""","

'* The end of the source file (pinmux setting).
#DEFINE ALL_END _
  !"\n    fragment@2 {" _
  !"\n      target = <&pruss>;" _
  !"\n      __overlay__ {" _
  !"\n          status = ""okay"";"_
  !"\n      };" _
  !"\n    };" _
  !"\n  };"


'* The start of fragment0 in the source file (pinmux settings).
#DEFINE FRAG0_START _
  !"\n      ""pruss"";" _
  !"\n" _
  !"\n    fragment@0 {" _
  !"\n      target = <&am33xx_pinmux>;" _
  !"\n      __overlay__ {"

'* The end of fragment0 in the source file (pinmux settings).
#DEFINE FRAG0_END _
  !"\n      };" _
  !"\n    };"


'* The start of fragment1 in the source file (pinmux_helper).
#DEFINE FRAG1_START _
  !"\n    fragment@1 {" _
  !"\n      target = <&ocp>;" _
  !"\n      __overlay__ {"

'* The end of fragment1 in the source file (pinmux_helper).
#DEFINE FRAG1_END _
  !"\n      };" _
  !"\n    };"


'* Enumerators for pin modes.
ENUM
  NP = &b001000 '*< no resistor connected
  PU = &b010000 '*< pullup resistor connected
  PD = &b000000 '*< pulldown resistor connected
  RX = &b100000 '*< input receiver enabled

  _O_ = NP      '*< setting for an output pin.
  _I_ = RX + PD '*< setting for an input pin.
  I_O = RX + NP '*< setting for an in-/out-put pin.
  IOD = RX + NP '*< setting for an in-/out-put pin.

  TMRi = _I_ '*< setting for a TIMER input pin.
  TMRo = _O_ '*< setting for a TIMER output pin.
  CAPi = _I_ '*< setting for a CAP input pin.
  CAPo = _O_ '*< setting for a CAP output pin.
  QEPi = _I_ '*< setting for a QEP input pin.
  QEPo = _O_ '*< setting for a QEP output pin (sync).
  PWMi = _I_ '*< setting for a PWM input pin (sync).
  PWMo = _O_ '*< setting for a PWM output pin.
END ENUM

'* The default settings for the GPIO modes.
#DEFINE GPIO_DEF CHR( _
    7 + RX + NP _ ' input, open (no resistor)
  , 7 + RX + PU _ ' input, pullup resistor
  , 7 + RX + PD _ ' input, pulldown resistor
  , 7 + NP  _     ' output (no resistor)
  )

'* Macro to delete the pin configuration for a pin set array.
#DEFINE PIN_DEL(_A_) FOR i AS LONG = 0 TO UBOUND(_A_) : M(_A_(i)) = "" : NEXT

'* The array to be filled with modus settings for all pins.
DIM SHARED AS STRING M(109)



/'* \brief Create lines for fragment0 for all settings of a pin.
\param I The index (ball number) of the pin in global array M.
\returns A string containing several lines with pin settings for fragment0.

FIXME

'/
FUNCTION f0custom(BYVAL I AS UBYTE) AS STRING
  VAR x = HEX(ASC(M(I)), 2) _
   , b4 = HEX(I * 4, 3)

  SELECT CASE AS CONST I
  CASE  89 : b4 = "1A0 0x2F  0x" & b4
  CASE 104 : b4 = "164 0x2F  0x" & b4
  CASE 106 : b4 = "1B4 0x2F  0x" & b4
  CASE 108 : b4 = "1A8 0x2F  0x" & b4
  END SELECT
  RETURN !"\n          0x" & b4 & " 0x" & x
END FUNCTION


/'* \brief Create lines for fragment0 for all settings of a pin.
\param I The index (ball number) of the pin in global array M.
\returns A string containing several lines with pin settings for fragment0.

FIXME

'/
FUNCTION f0entry(BYVAL I AS UBYTE) AS STRING
  VAR r = "" _
    , b0 = HEX(I, 2) _
    , b4 = HEX(I * 4, 3)

  SELECT CASE AS CONST I
  CASE  89 : b4 = "1A0 0x2F  0x" & b4
  CASE 104 : b4 = "164 0x2F  0x" & b4
  CASE 106 : b4 = "1B4 0x2F  0x" & b4
  CASE 108 : b4 = "1A8 0x2F  0x" & b4
  END SELECT

  FOR j AS LONG = 0 TO LEN(M(I)) - 1
    VAR x = HEX(M(I)[j], 2) _
      , n = b0 & "_" & x
    r &= _
      !"\n        B" & n & ": " & n _
      & " {pinctrl-single,pins = <0x" & b4 & " 0x" & x & ">;};"
  NEXT : RETURN r
END FUNCTION


/'* \brief Create lines for fragment1 for all settings of a pin.
\param I The index (ball number) of the pin in global array M.
\returns A string containing an entry with pin settings for fragment1.

FIXME

\note The maximum number of configurations for a pin is 36.

'/
FUNCTION f1entry(BYVAL I AS UBYTE) AS STRING
  VAR _
    j = 0 _
  , tn = HEX(I, 2) _
  , t0 = _
    !"\n        " & PMUX_NAME & tn & " {" _
    !"\n          compatible = ""bone-pinmux-helper"";" _
    !"\n          status = ""okay"";" _
    !"\n          pinctrl-names = " _
  , t1 = _
    !"\n        };" _
  , l = "" _
  , n = ""

  FOR j = 0 TO LEN(M(I)) - 2
    VAR t = HEX(M(I)[j], 2)
    n &= """x" & t & """, "
    l &= !"\n          pinctrl-" & j & " = <&B" & tn & "_" & t & ">;"
  NEXT
  VAR t = HEX(M(I)[j], 2)
  n &= """x" & t & """;"
  l &= !"\n          pinctrl-" & j & " = <&B" & tn & "_" & t & ">;"
  RETURN t0 & n & l & t1
END FUNCTION


/'* \brief Get header pin connected to CPU ball.
\param Ball The CPU ball number.
\returns A string pointer (don't free it) on success (otherwise zero).
\since 0.2

This function creates a text description of the header pin connected to
a CPU ball. The returned string is owned by this function and must not
be freed.

When the CPU ball is not connected to a header pin, this function
returns 0 (zero).

'/
FUNCTION nameBall CDECL(BYVAL Ball AS UBYTE) AS ZSTRING PTR
  SELECT CASE AS CONST Ball '                                  find name
  CASE   6 : RETURN @"P8.03"
  CASE   7 : RETURN @"P8.04"
  CASE   2 : RETURN @"P8.05"
  CASE   3 : RETURN @"P8.06"
  CASE  36 : RETURN @"P8.07"
  CASE  37 : RETURN @"P8.08"
  CASE  39 : RETURN @"P8.09"
  CASE  38 : RETURN @"P8.10"
  CASE  13 : RETURN @"P8.11"
  CASE  12 : RETURN @"P8.12"
  CASE   9 : RETURN @"P8.13"
  CASE  10 : RETURN @"P8.14"
  CASE  15 : RETURN @"P8.15"
  CASE  14 : RETURN @"P8.16"
  CASE  11 : RETURN @"P8.17"
  CASE  35 : RETURN @"P8.18"
  CASE   8 : RETURN @"P8.19"
  CASE  33 : RETURN @"P8.20"
  CASE  32 : RETURN @"P8.21"
  CASE   5 : RETURN @"P8.22"
  CASE   4 : RETURN @"P8.23"
  CASE   1 : RETURN @"P8.24"
  CASE   0 : RETURN @"P8.25"
  CASE  31 : RETURN @"P8.26"
  CASE  56 : RETURN @"P8.27"
  CASE  58 : RETURN @"P8.28"
  CASE  57 : RETURN @"P8.29"
  CASE  59 : RETURN @"P8.30"
  CASE  54 : RETURN @"P8.31"
  CASE  55 : RETURN @"P8.32"
  CASE  53 : RETURN @"P8.33"
  CASE  51 : RETURN @"P8.34"
  CASE  52 : RETURN @"P8.35"
  CASE  50 : RETURN @"P8.36"
  CASE  48 : RETURN @"P8.37"
  CASE  49 : RETURN @"P8.38"
  CASE  46 : RETURN @"P8.39"
  CASE  47 : RETURN @"P8.40"
  CASE  44 : RETURN @"P8.41"
  CASE  45 : RETURN @"P8.42"
  CASE  42 : RETURN @"P8.43"
  CASE  43 : RETURN @"P8.44"
  CASE  40 : RETURN @"P8.45"
  CASE  41 : RETURN @"P8.46"
  CASE  28 : RETURN @"P9.11"
  CASE  30 : RETURN @"P9.12"
  CASE  29 : RETURN @"P9.13"
  CASE  18 : RETURN @"P9.14"
  CASE  16 : RETURN @"P9.15"
  CASE  19 : RETURN @"P9.16"
  CASE  87 : RETURN @"P9.17"
  CASE  86 : RETURN @"P9.18"
  CASE  95 : RETURN @"P9.19"
  CASE  94 : RETURN @"P9.20"
  CASE  85 : RETURN @"P9.21"
  CASE  84 : RETURN @"P9.22"
  CASE  17 : RETURN @"P9.23"
  CASE  97 : RETURN @"P9.24"
  CASE 107 : RETURN @"P9.25"
  CASE  96 : RETURN @"P9.26"
  CASE 105 : RETURN @"P9.27"
  CASE 103 : RETURN @"P9.28"
  CASE 101 : RETURN @"P9.29"
  CASE 102 : RETURN @"P9.30"
  CASE 100 : RETURN @"P9.31"
  CASE 109 : RETURN @"P9.41"
  CASE  89 : RETURN @"P9.42"
  END SELECT : RETURN 0
END FUNCTION
